cmd/kubeadm/app/util/etcd/etcd.go

```diff
@@ -421,29 +421,19 @@ func (c *Client) addMember(name string, peerAddrs string, isLearner bool) ([]Mem
 	var (
 		lastError   error
 		respMembers []*etcdserverpb.Member
-		learnerID   uint64
 		resp        *clientv3.MemberAddResponse
 	)
 	err = wait.PollUntilContextTimeout(context.Background(), constants.EtcdAPICallRetryInterval, constants.EtcdAPICallTimeout,
 		true, func(_ context.Context) (bool, error) {
 			ctx, cancel := context.WithTimeout(context.Background(), etcdTimeout)
 			defer cancel()
 			if isLearner {
-				// if learnerID is set, it means the etcd member is already added successfully.
-				if learnerID == 0 {
-					klog.V(1).Info("[etcd] Adding etcd member as learner")
-					resp, err = cli.MemberAddAsLearner(ctx, []string{peerAddrs})
-					if err != nil {
-						lastError = err
-						return false, nil
-					}
-					learnerID = resp.Member.ID
-				}
-				respMembers = resp.Members
-				return true, nil
+				klog.V(1).Infof("[etcd] Adding etcd member %q as learner", peerAddrs)
+				resp, err = cli.MemberAddAsLearner(ctx, []string{peerAddrs})
+			} else {
+				klog.V(1).Infof("[etcd] Adding etcd member %q", peerAddrs)
+				resp, err = cli.MemberAdd(ctx, []string{peerAddrs})
```

Comment:
i don't think that this would fix the linked issue. i.e. just checking for `if errors.Is(err, rpctypes.ErrPeerURLExist) {` after the add does not suffice:

here is my understanding:
- ListMembers inside the Poll
- if the URL does not exist call Add

same for Remove:
- here is my understanding:
- if the URL exists call Remove

Reply:
Emm... But I think if the first try failed because of the network timeout, the second try should just return the existing error, is there any other scenario? Can we trust the etcd `ErrPeerURLExist` error type when network is OK?

Reply:
Before this fix, we did not check whether `MemberAddAsLearner` failure was caused by the `Exist` error (it just quick returned with an error), which might be the cause of the problem.

Reply:
that's a good point. does this error check after exist in older versions in support skew too?

i think the discussed scenario was when the member add continuously returns "context deadline" and never returns the "peer exist" error. but a member was added already. according to the report  i saw downstream they never hit the "peer exists" error during add. then if the user tries to add the same CP node it may fail with existing member.

the list before add / remove seems a bit too much...and also what does it mean if the list members calls has "context deadline" then the member add should fail. probably yes.

in any case, we can add this:
```
		// inside the poll for add / remove
			resp, err = cli.MemberList(ctx)
			// check if member exists
			// do add / remove here
```
to avoid these cases of the return err from add/remove not being trusable.
WDYT?

Reply:
> Before this fix, we did not check whether `MemberAddAsLearner` failure was caused by the `Exist` error (it just quick returned with an error, didn't check the error type), which might be the cause of the problem.

i think the learnerID is not needed, yes.
but the case i describe in the above comment i think is different.

Reply:
Updated @neolit123 @pacoxu